09-intermediate-rails/testing-auth.md

```diff
@@ -52,151 +73,284 @@ class ActiveSupport::TestCase
     OmniAuth.config.test_mode = true
-
-    # The mock_auth configuration allows you to set per-provider (or default) authentication
-    # hashes to return during testing.
-    OmniAuth.config.mock_auth[:github] = OmniAuth::AuthHash.new({
-      provider: 'github', uid: '123545', info: { email: "a@b.com", name: "Ada" }
-      })
   end
 end
 ```
 
-The above code will populate the **default** OmniAuth request data hash with the information that you have provided here. This is especially useful and necessary when you are testing controller actions that may require a user to be logged in. This is mock data which should match what we may expect to see from the _real_ `auth_hash` we'll get back from GitHub. This is setting up the **default** data and nothing more. We'll need to _use_ this data in our tests in the next step.
+The `setup` method defined here will be run once, before all tests. All of our `describe` blocks will "inherit" from this class, so the setup code will be applied to them too.
 
-#### Session Tests
+#### Define Test Fixtures
 
-First, let's create a method which will allow us to log in using the default OmniAuth data that we just configured.
+These test fixtures are very similar to those we've made in the past. The only difference is in how we'll use them, not in how they are defined. Make sure the field names match the schema for your User model (this may differ slightly between Stacks and Queues).
+
+<!-- TODO: Add yml highlighting before finishing -->
+```
+ada:
+  oauth_provider: github
+  oauth_uid: 12345
+  email: ada@adadevelopersacademy.org
+  username: countess_ada
+
+grace:
+  oauth_provider: github
+  oauth_uid: 13371337
+  email: grace@hooper.net
+  username: graceful_hoops
+```
+
+#### Logging In
+
+Next, we need to use this information to log in. We'll start by testing the auth callback itself, which will allow us to exercise this functionality in isolation.
+
+**Question:** What interesting test cases can you think of for the auth callback? Hint: there are at least 3.
+
+##### Building the Auth Hash
+
+In order to log in a user, we'll need to provide data to our controller in the exact same format as GitHub does. To make this process easy, we'll create an instance method for the User model, `User#mock_auth_hash`. This method will be almost the opposite of the `from_github` method we defined when we set up OAuth in the first place.
 
 ```ruby
-# sessions_controller_test.rb
-describe SessionsController do
-  def login
-    open_session do |sess|
-      sess.get auth_github_callback_path, env: { # using our app's callback route
-        'omniauth.auth' => OmniAuth.config.mock_auth[:github] # using the configuration we added above
+# app/models/user.rb
+class User < ApplicationRecord
+  # ...
+  # Whatever was here before (relations, validations, etc.)
+  # ...
+
+  # Test helper method to generate a mock auth hash
+  # for fixture data
+  def mock_auth_hash
```

Comment:
It just seems off to me to modify the model with test-code.  Test code, I feel should go into our testing files, not the models or controllers.

Maybe instead in the test_helper file.

```ruby
def mock_auth_hash(user)
  return {
    email: user.email,
    uid = user.oauth_id,
    info: {
      email: user.email,
      nickname: user.username
    }
  }
end
```